omnigibson/tasks/behavior_task.py

```diff
@@ -544,405 +286,63 @@ def assign_object_scope_with_cache(self, env):
-        # E.g. after sampling Filled.set_value(True), Filled.get_value() will become True only after one step
-        og.sim.step()
-
-    def sample(self, env, validate_goal=False):
-        """
-        Run sampling for this BEHAVIOR task
-
-        Args:
-            env (Environment): Current active environment instance
-            validate_goal (bool): Whether the goal should be validated or not
-
-        Returns:
-            2-tuple:
-                - bool: Whether sampling was successful or not
-                - None or str: None if successful, otherwise the associated error message
-        """
-        # Auto-initialize all sampleable objects
-        with og.sim.playing():
-            env.scene.reset()
-
-            error_msg = self.group_initial_conditions()
-            if error_msg:
-                log.error(error_msg)
-                return False, error_msg
-
-            error_msg = self.sample_initial_conditions()
-            if error_msg:
-                log.error(error_msg)
-                return False, error_msg
-
-            if validate_goal:
-                error_msg = self.sample_goal_conditions()
-                if error_msg:
-                    log.error(error_msg)
-                    return False, error_msg
-
-            error_msg = self.sample_initial_conditions_final()
-            if error_msg:
-                log.error(error_msg)
-                return False, error_msg
-
-            env.scene.update_initial_state()
-
-        return True, None
+            assert entity is not None, f"Could not assign valid BDDLEntity for {obj_inst}!"
+            self.object_scope[obj_inst] = entity
 
     def _get_obs(self, env):
         low_dim_obs = dict()
-        low_dim_obs["robot_pos"] = np.array(env.robots[0].get_position())
-        low_dim_obs["robot_ori_cos"] = np.cos(env.robots[0].get_rpy())
-        low_dim_obs["robot_ori_sin"] = np.sin(env.robots[0].get_rpy())
 
         # Batch rpy calculations for much better efficiency
-        objs_rpy = T.quat2euler(np.array([v.states[Pose].get_value()[1] for v in self.object_scope.values()]))
+        objs_exist = {obj: obj.exists() for obj in self.object_scope.values() if not obj.is_system}
+        objs_rpy = T.quat2euler(np.array([obj.states[Pose].get_value()[1] if obj_exist else np.array([0, 0, 0, 1.0])
+                                          for obj, obj_exist in objs_exist.items()]))
+
+        # Always add agent info first
+        agent = self.get_agent(env=env)
+        low_dim_obs["robot_pos"] = agent.states[Pose].get_value()[1]
+        low_dim_obs["robot_ori_cos"] = np.cos(objs_rpy[self.agent_obj_idx])
+        low_dim_obs["robot_ori_sin"] = np.sin(objs_rpy[self.agent_obj_idx])
+
+        for idx, ((obj, obj_exist), obj_rpy) in enumerate(zip(objs_exist.items(), objs_rpy)):
+            # Skip if agent idx -- note: doesn't increment object!
+            if idx == self.agent_obj_idx:
```

Comment:
hmm is there a better way to handle this? this `self.agent_obj_idx` doesn't seem too clean. Can we just skip synset `agent.n.01`?

Reply:
fixed